### ISSUE TITLE
Custom widget access to form Data

### ISSUE THREAD

**Issue by zzph** [rjsf-team/react-jsonschema-form#4192]
### Prerequisites

- [X] I have searched the existing issues
- [X] I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [X] I have read the [documentation](https://rjsf-team.github.io/react-jsonschema-form/docs)
- [X] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net), [Codesandbox.io](https://codesandbox.io) or preferably a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### What theme are you using?

core

### Version

5.x

### Current Behavior

Inside my custom widget, it will not let me access the current formData. So I pass it into `formContext` but that breaks the form (nothing can be inputted).

### Expected Behavior

Inside my custom widget, I can access all the current values of the form

### Steps To Reproduce

1. Using react and rjsf (both on latest)
2. Using a custom widget (as implemented per docs)
3. I expect the `formData` to appear as a prop, but it does not, so I try passing it on formContext instead (as per issue [here](https://github.com/rjsf-team/react-jsonschema-form/issues/1541)
4. Using that formContext freezes the entire form and nothing can be inputted.

I simplified my code as per below:

`import { useState } from 'react'
import Form from '@rjsf/core';
import validator from '@rjsf/validator-ajv8';

const TestCmp = (props) => {
  const { formContext } = props;
  // Use formData as needed
  // ...
  return <div> TEST {JSON.stringify(formContext)} </div>
};

const uiSchema = {
  name: {
    'ui:widget': 'TestCmp'
  }
};
const schema = {
  "type": "object",
  "properties": {
    "name": {
      "type": "string"
    },
    "age": {
      "type": "number"
    }
  }
}

const widgets = {
  TestCmp
};

export default function Create() {
  const [formDataPublic, setFormDataPublic ] = useState({})
  return <Form
        schema={schema}
        validator={validator}
        uiSchema={uiSchema}
        widgets={widgets}
        onChange={({formData})=> setFormDataPublic(formData)}
        formContext={{formDataPublic}}
      />
}`

### Environment

```markdown
- OS: MacOS
- Node: 20.11
- npm: 10.2.4
```

**Comment by zzph** [comment]
Any ideas on this? I'm blocked from using this awesome library because of it

**Comment by heath-freenome** [comment]
The widget has access to the `formData` at its current level (in your example the widget will get the value for `name`) via the `value` prop. Is there a reason why you need access to all of the `formData` within a single widget?

**Comment by zzph** [comment]
Thanks for the reply.

In this case, the field would “auto suggest” a value based on other fields
(which can also be corrected by the user if not correct).

Is that clear? Is there a way to achieve it?

**Comment by yarnball** [comment]
Yes I've debugged, and it does seem like you're saying- it re-renders the entire component everytime 'formContext' changes. So this is an issue.

Can you suggest a work around?

**Comment by zzph** [comment]
Yes I've debugged, and it does seem like you're saying- it re-renders the entire component everytime 'formContext' changes. So this is an issue.

Can you suggest a work around?

**Comment by heath-freenome** [comment]
Maybe you can process the formData on change in a way that you transform what you put into the formContext is only what you need to provide the `auto-suggest`? Meaning if your component looks at the value of `age` to auto-suggest the value, perhaps precompute the suggestion OUTSIDE of the component and pass in the suggestion rather than the formData?

**Comment by hawkzey** [comment]
I have a similar issue to this, I am trying to access the value of a property inside of a widget for a different property.

I tried having a controlled state outside my form and the on change to update that, i could then pass this property thought like you suggest but this causes a re render loop :/ ( the onChange causes this).

@zzph did you make any progress on this?

**Comment by mateuszbaldyga** [comment]
In order to make it work you need to pass `formData` prop  to a `<Form />` component:

```js
const [formData, setFormData] = useState({});

return (
  <Form
    formData={formData}
    onChange={({ formData }) => {
      setFormData(formData);
    }}
    formContext={{ formData }}
  />
)
```

**Comment by e-senturk** [comment]
@mateuszbaldyga This approach is unhelpful. This use case breaks the form validation feature because it causes rerendering. It also works extremely laggy. We need an uncontrolled approach to solve this issue.

**Comment by Fronix** [comment]
> [@mateuszbaldyga](https://github.com/mateuszbaldyga) This approach is unhelpful. This use case breaks the form validation feature because it causes rerendering. It also works extremely laggy. We need an uncontrolled approach to solve this issue.

Indeed, we've been using the context/state approach but we've had to make some rather ugly fixes to combat the extrem lag of re-rendering everything everytime formdata changes.
